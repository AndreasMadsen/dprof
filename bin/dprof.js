#!/bin/env node

var endpoint = require('endpoint');
var extend = require('util-extend');

process.stdin.pipe(endpoint(function (err, dprof) {
  if (err) throw err;
  dprof = JSON.parse(dprof);

  var content = new Node(dprof.root, dprof.total, 1, new State(), new Root());

  var svg = new Element('svg').attr({ xmlns: 'http://www.w3.org/2000/svg', width: 1000 });

  process.stdout.write(svg.start() + '\n');
  process.stdout.write(content.stringify() + '\n');
  process.stdout.write(svg.end());
}));

function Root() {
  this.top = 0;
}

function State() {
  this.y = 0;
}

function Node(node, total, depth, state, parent) {
  this.name = node.name;
  this.init = Math.floor(node.init / total * 1000);
  this.before = Math.floor((node.init + node.before) / total * 1000);
  this.after = Math.floor((node.init + node.after) / total * 1000);
  this.stack = node.stack;

  this.depth = depth;
  this.state = state;
  this.parent = parent;
  this.top = state.y;

  this.elems = [];
  this.elems.push(new Element('rect').attr({
    x: this.init,
    y: this.top,
    width: this.before - this.init,
    height: 20,
    fill: 'SteelBlue',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('rect').attr({
    x: this.before,
    y: this.top,
    width: this.after - this.before,
    height: 20,
    fill: 'IndianRed',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('line').attr({
    x1: this.init,
    y1: this.parent.top + 20,
    x2: this.init,
    y2: this.top + 20,
    style: 'stroke: #000; stroke-width: 2px'
  }));

  this.state.y += 20;
  this.children = node.children.map(function (child) {
    return new Node(child, total, depth + 1, state, this);
  }, this);
}

Node.prototype._indent = function () {
  var s = '';
  for (var i = 0; i < this.depth; i++) {
    s += '  ';
  }
  return s;
};

Node.prototype.stringify = function () {
  var content = [];
  this.elems.forEach(function (elem) {
    content.push(this._indent() + elem.singleton());
  }, this);
  this.children.forEach(function (child) {
    content.push(child.stringify());
  }, this);

  return content.join('\n');
};

function Element(name) {
  this._name = name;
  this._attr = {};
}

Element.prototype.attr = function (attr) {
  extend(this._attr, attr);
  return this;
};

Element.prototype.start = function () {
  var start = '<' + this._name;
  Object.keys(this._attr).forEach(function (k) {
    start += ' ' + k + '="' + this._attr[k] + '"';
  }, this);
  start += '>';
  return start;
};

Element.prototype.end = function () {
  return '</' + this._name + '>';
};

Element.prototype.singleton = function () {
  return this.start() + this.end();
};
