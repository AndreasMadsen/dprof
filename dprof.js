'use strict';

const providers = process.binding('async_wrap').Providers;
const asyncWrap = require('./async_wrap.js');
const zlib = require('zlib');
const fs = require('fs');

const version = require('./package.json').version;
const processStart = process.hrtime();

// Change the default stackTraceLimit if it haven't allready been overwritten
if (process.execArgv.indexOf('--stack_trace_limit') === -1 && Error.stackTraceLimit === 10) {
  Error.stackTraceLimit = 8;
}

//
// Define node class
//

function Site(site) {
  this.description = site.toString();
  this.filename = site.getFileName();
  this.line = site.getLineNumber();
  this.collum = site.getColumnNumber();
}

function timestamp() {
  const t = process.hrtime(processStart);
  return t[0] * 1e9 + t[1];
}

function Node(name, stack) {
  this.name = name;
  this._init = timestamp();
  this._before = [];
  this._after = [];
  this.children = [];
  this.stack = stack.map(function (site) {
    return new Site(site);
  });
}

Node.prototype.add = function (handle) {
  const node = new Node(handle.constructor.name, asyncWrap.stackTrace(3));
  this.children.push(node);
  return node;
};

Node.prototype.before = function () {
  this._before.push(timestamp());
};

Node.prototype.after = function () {
  this._after.push(timestamp());
};

Node.prototype.toJSON = function () {
  return {
    name: this.name,
    init: this._init,
    before: this._before,
    after: this._after,
    stack: this.stack,
    children: this.children
  };
};

Node.prototype.rootIntialize = function () {
  this._init = 0;
  this._before.push(0);
};

//
// Setup hooks
//

asyncWrap.setup(asyncInit, asyncBefore, asyncAfter);

const root = new Node('root', asyncWrap.stackTrace(2));
      root.rootIntialize();
let state = root;

function asyncInit(provider, parent) {
  if (provider === providers.TIMERWRAP) {
    this._dprofIgnore = true;
    return;
  }

  if (parent) {
    this._dprofState = parent._dprofState.add(this, provider, parent);
  } else {
    this._dprofState = state.add(this, provider, parent);
  }
}

function asyncBefore() {
  if (this._dprofIgnore) return;

  this._dprofState.before();
  state = this._dprofState;
}

function asyncAfter() {
  if (this._dprofIgnore) return;
  
  this._dprofState.after();
  state = root;
}

// The root job is done when process.nextTick is called
asyncWrap.disable();
process.nextTick(function () {
  root.after();
});
asyncWrap.enable();

//
// Print result
//

process.on('exit', function () {
  // even though zlib is sync, it still fires async_wrap events,
  // so disable asyncWrap just to be sure.
  asyncWrap.disable();

  const data = {
    'total': timestamp(),
    'version': version,
    'root': root
  };

  if (process.env.NODE_DPROF_DEBUG) {
    fs.writeFileSync('./dprof.json', JSON.stringify(data, null, 1));
  } else {
    fs.writeFileSync('./dprof.json.gz', zlib.gzipSync(JSON.stringify(data)));
  }
});
