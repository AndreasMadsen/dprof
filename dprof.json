{
 "name": "root",
 "init": 0,
 "before": 0,
 "after": 4319266,
 "children": [
  {
   "name": "Timer",
   "init": 2044586,
   "before": 548667,
   "after": 1409916,
   "children": [
    {
     "name": "FSReqWrap",
     "init": 3039113,
     "before": 458225,
     "after": 788615,
     "children": [
      {
       "name": "FSReqWrap",
       "init": 3776193,
       "before": 164991,
       "after": 213630,
       "children": []
      },
      {
       "name": "FSReqWrap",
       "init": 3803875,
       "before": 188243,
       "after": 268189,
       "children": [
        {
         "name": "FSReqWrap",
         "init": 4063866,
         "before": 39850,
         "after": 74510,
         "children": []
        }
       ]
      }
     ]
    },
    {
     "name": "FSReqWrap",
     "init": 3279865,
     "before": 623396,
     "after": 650968,
     "children": []
    }
   ]
  }
 ]
}