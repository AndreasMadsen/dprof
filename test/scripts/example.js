'use strict';
require('../../dprof.js');

const fs = require('fs');

setTimeout(function () {

  fs.open(__filename, 'r', function (err1, fd) {
    if (err1) throw err1;

    let count = 0;
    const a = new Buffer(10);
    fs.read(fd, a, 0, 10, 0, function (err2) {
      if (err2) throw err2;

      count += 1;
      if (count === 2) close();
    });

    const b = new Buffer(10);
    fs.read(fd, b, 0, 10, 10, function (err2) {
      if (err2) throw err2;

      count += 1;
      if (count === 2) close();
    });

    function close() {
      fs.close(fd, function (err2) {
        if (err2) throw err2;
      });
    }
  });

  fs.readdir(__dirname, function (err) {
    if (err) throw err;
  });
}, 1);
