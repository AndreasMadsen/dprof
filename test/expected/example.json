{
 "stdout": "",
 "stderr": "",
 "dump": {
  "total": 7593346,
  "version": "0.14.0",
  "root": {
   "name": "root",
   "init": 0,
   "destroy": 4346887,
   "before": [
    0
   ],
   "after": [
    4316288
   ],
   "stack": [
    {
     "description": "Module._compile (module.js:541:32)",
     "filename": "module.js",
     "line": 541,
     "collum": 32
    },
    {
     "description": "Object.Module._extensions..js (module.js:550:10)",
     "filename": "module.js",
     "line": 550,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:458:32)",
     "filename": "module.js",
     "line": 458,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:417:12)",
     "filename": "module.js",
     "line": 417,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:409:3)",
     "filename": "module.js",
     "line": 409,
     "collum": 3
    },
    {
     "description": "Module.require (module.js:468:17)",
     "filename": "module.js",
     "line": 468,
     "collum": 17
    },
    {
     "description": "require (internal/module.js:20:19)",
     "filename": "internal/module.js",
     "line": 20,
     "collum": 19
    },
    {
     "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/example.js:2:1)",
     "filename": "/test/scripts/example.js",
     "line": 2,
     "collum": 1
    },
    {
     "description": "Module._compile (module.js:541:32)",
     "filename": "module.js",
     "line": 541,
     "collum": 32
    },
    {
     "description": "Object.Module._extensions..js (module.js:550:10)",
     "filename": "module.js",
     "line": 550,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:458:32)",
     "filename": "module.js",
     "line": 458,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:417:12)",
     "filename": "module.js",
     "line": 417,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:409:3)",
     "filename": "module.js",
     "line": 409,
     "collum": 3
    },
    {
     "description": "Module.runMain (module.js:575:10)",
     "filename": "module.js",
     "line": 575,
     "collum": 10
    },
    {
     "description": "run (bootstrap_node.js:352:7)",
     "filename": "bootstrap_node.js",
     "line": 352,
     "collum": 7
    },
    {
     "description": "startup (bootstrap_node.js:144:9)",
     "filename": "bootstrap_node.js",
     "line": 144,
     "collum": 9
    },
    {
     "description": "bootstrap_node.js:467:3",
     "filename": "bootstrap_node.js",
     "line": 467,
     "collum": 3
    }
   ],
   "children": [
    {
     "name": "TimeoutWrap",
     "init": 3234326,
     "destroy": 6324392,
     "before": [
      4910358
     ],
     "after": [
      6090412
     ],
     "stack": [
      {
       "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/example.js:6:1)",
       "filename": "/test/scripts/example.js",
       "line": 6,
       "collum": 1
      },
      {
       "description": "Module._compile (module.js:541:32)",
       "filename": "module.js",
       "line": 541,
       "collum": 32
      },
      {
       "description": "Object.Module._extensions..js (module.js:550:10)",
       "filename": "module.js",
       "line": 550,
       "collum": 10
      },
      {
       "description": "Module.load (module.js:458:32)",
       "filename": "module.js",
       "line": 458,
       "collum": 32
      },
      {
       "description": "tryModuleLoad (module.js:417:12)",
       "filename": "module.js",
       "line": 417,
       "collum": 12
      },
      {
       "description": "Function.Module._load (module.js:409:3)",
       "filename": "module.js",
       "line": 409,
       "collum": 3
      },
      {
       "description": "Module.runMain (module.js:575:10)",
       "filename": "module.js",
       "line": 575,
       "collum": 10
      },
      {
       "description": "run (bootstrap_node.js:352:7)",
       "filename": "bootstrap_node.js",
       "line": 352,
       "collum": 7
      },
      {
       "description": "startup (bootstrap_node.js:144:9)",
       "filename": "bootstrap_node.js",
       "line": 144,
       "collum": 9
      },
      {
       "description": "bootstrap_node.js:467:3",
       "filename": "bootstrap_node.js",
       "line": 467,
       "collum": 3
      }
     ],
     "children": [
      {
       "name": "FSReqWrap",
       "init": 5137186,
       "destroy": 6927139,
       "before": [
        6418801
       ],
       "after": [
        6911356
       ],
       "stack": [
        {
         "description": "Object.fs.open (fs.js:631:11)",
         "filename": "fs.js",
         "line": 631,
         "collum": 11
        },
        {
         "description": "Timeout.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/example.js:8:6)",
         "filename": "/test/scripts/example.js",
         "line": 8,
         "collum": 6
        },
        {
         "description": "tryOnTimeout (timers.js:228:11)",
         "filename": "timers.js",
         "line": 228,
         "collum": 11
        },
        {
         "description": "Timer.listOnTimeout (timers.js:202:5)",
         "filename": "timers.js",
         "line": 202,
         "collum": 5
        }
       ],
       "children": [
        {
         "name": "FSReqWrap",
         "init": 6736433,
         "destroy": 7184487,
         "before": [
          7124834
         ],
         "after": [
          7180561
         ],
         "stack": [
          {
           "description": "Object.fs.read (fs.js:687:11)",
           "filename": "fs.js",
           "line": 687,
           "collum": 11
          },
          {
           "description": "/Users/Andreas/Sites/node_modules/dprof/test/scripts/example.js:13:8",
           "filename": "/test/scripts/example.js",
           "line": 13,
           "collum": 8
          },
          {
           "description": "FSReqWrap.oncomplete (fs.js:123:15)",
           "filename": "fs.js",
           "line": 123,
           "collum": 15
          }
         ],
         "children": []
        },
        {
         "name": "FSReqWrap",
         "init": 6858905,
         "destroy": 7374832,
         "before": [
          7188483
         ],
         "after": [
          7370699
         ],
         "stack": [
          {
           "description": "Object.fs.read (fs.js:687:11)",
           "filename": "fs.js",
           "line": 687,
           "collum": 11
          },
          {
           "description": "/Users/Andreas/Sites/node_modules/dprof/test/scripts/example.js:21:8",
           "filename": "/test/scripts/example.js",
           "line": 21,
           "collum": 8
          },
          {
           "description": "FSReqWrap.oncomplete (fs.js:123:15)",
           "filename": "fs.js",
           "line": 123,
           "collum": 15
          }
         ],
         "children": [
          {
           "name": "FSReqWrap",
           "init": 7307910,
           "destroy": 7453932,
           "before": [
            7414926
           ],
           "after": [
            7449866
           ],
           "stack": [
            {
             "description": "Object.fs.close (fs.js:605:11)",
             "filename": "fs.js",
             "line": 605,
             "collum": 11
            },
            {
             "description": "close (/Users/Andreas/Sites/node_modules/dprof/test/scripts/example.js:29:10)",
             "filename": "/test/scripts/example.js",
             "line": 29,
             "collum": 10
            },
            {
             "description": "/Users/Andreas/Sites/node_modules/dprof/test/scripts/example.js:25:24",
             "filename": "/test/scripts/example.js",
             "line": 25,
             "collum": 24
            },
            {
             "description": "FSReqWrap.wrapper [as oncomplete] (fs.js:681:17)",
             "filename": "fs.js",
             "line": 681,
             "collum": 17
            }
           ],
           "children": []
          }
         ]
        }
       ]
      },
      {
       "name": "FSReqWrap",
       "init": 5775829,
       "destroy": 7027928,
       "before": [
        6995428
       ],
       "after": [
        7023671
       ],
       "stack": [
        {
         "description": "Object.fs.readdir (fs.js:941:11)",
         "filename": "fs.js",
         "line": 941,
         "collum": 11
        },
        {
         "description": "Timeout.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/example.js:35:6)",
         "filename": "/test/scripts/example.js",
         "line": 35,
         "collum": 6
        },
        {
         "description": "tryOnTimeout (timers.js:228:11)",
         "filename": "timers.js",
         "line": 228,
         "collum": 11
        },
        {
         "description": "Timer.listOnTimeout (timers.js:202:5)",
         "filename": "timers.js",
         "line": 202,
         "collum": 5
        }
       ],
       "children": []
      }
     ]
    }
   ]
  }
 }
}